src/app/formFields/getCustomFormFieldsValidationSchema.ts

```diff
@@ -29,7 +28,7 @@ export default memoize(function getAddressCustomFieldsValidationSchema({
             formFields
                 .filter(({ custom }) => !!custom)
                 .reduce((schema, { name, label, required, fieldType, type, min, max }) => {
-                    const requiredErrorMessage = translate(`${ERROR_KEYS.custom}_required_error`, { label });
+                    const requiredErrorMessage = translate('address.custom_required_error', { label });
```

Comment:
The translation keys are still tied to the address namespace. If we want this to be a generic function, we should consider lifting the translation keys up and passing them into the function instead. For example, `SingleShippingForm` (where it is used at the moment) can pass in a map of translation keys specific to addresses. Alternatively, you can pass in a decorated `translate` function that returns address-specific error messages.

Reply:
I understand that currently, the translation key is `address.custom_required_error` however,I don't it's desirable to customize "custom form field required" in different ways for the address form and the account creation form. I opted to leave it in the address namespace, instead of moving to something more generic like `form.custom_required_error` as that would break existing translations. Thoughts?

Reply:
I actually thought about suggesting to take the simple route of just renaming the namespace but it'd definitely break existing translations. That's why I suggested to just lift the keys out of the function and pass them in instead. I made the suggestion because this component is no longer for a custom field for addresses. Instead it is for a custom field that can be applied to various contexts (address, account creation etc...), even if ultimately they end up with the same text. Therefore I believe it's best if `address` reference is removed from it.

Separately, we have to consider if we need to define a new key, e.g.: `customer.custom_required_error`, thereby giving giving merchants the ability to specify a different label for account creation. I agree it's probably not _strictly_ necessary because most likely they end up with the same text. But on the other hand, it does make it much clearer that the label belongs to a different section and describes a different thing.

Reply:
After thinking about how moving the keys up would end up looking like, I'm not convinced in would be a very clean design, as I would have to expose address-specific keys in the signature of `getFormFieldsValidationSchema`:

```js
// passed translation keys
{
  firstNameRequired,
  lastNameRequired,
  companyRequired,
  // rest of address specific fields
  // ...,
  customRequired,
  customMin
  customMax
  invalidCharacters
}
```

or even something a bit more complex but generic like 2 new params, `keys` and `overrides`
```js
// base translation keys
{
  required,
  min
  max
  invalidCharacters
}
// override keys
{
 firstNameRequired,
  lastNameRequired,
  companyRequired,
  // ... rest of address specific fields
}
```

I had instead created a new namespace `fields` that if present, takes precedence over the "legacy" way of translating form field error messages. Thoughts?

Reply:
I guess we have to take a step back a little bit. The interface of `ERROR_KEYS` (in master) is `{ [fieldName: string]: string }`. If you extract the translation keys out and pass them in, the interface is still `{ [fieldName: string]: string }`. You won't be passing translation keys specifically for `address`.

If you don't want to pass in translation keys, what you can do is to pass a `translate` function into the component, whose interface can look something like this `translate(key: string): string`. Similarly, it's not tied to `address`. So in both cases, you won't hit the problem you're describing and we can achieve the generalisation we want to achieve.

Reply:
I ended up moving address-specific keys up, but still left what I believe are general errors that should be shared. Let me know what you think. For compatibility with existing translations, I had to leave the old key. Or I could keep the general keys under the "address" namespace. Thoughts?

Reply:
Cool, I think you're almost there. Have you tried passing in a `translate` function instead, it seems to me that it's more fitting for your situation, e.g.:

```ts
const addressKeys = {
    address1: 'address.address_line_1',
    // etc...
};

// `address/translateAddressFieldError.ts`
function translateAddressFieldError(language: Language): string {
  return (fieldName: string, errorType: string) => {
    if (errorType === 'required') {
      return language.translate(`${addressKeys[fieldName]}_required_error`);
    }

    if (errorType === 'max') {
      return language.translate('address.max_error');
    }

    // etc...
  }
}
```

Then you can pass it into `getFormFieldsValidationSchema`

```ts
getFormFieldsValidationSchema({ translate: translateAddressFieldError })

function getFormFieldsValidationSchema({ translate }) {
  // etc...

  if (maxValue !== undefined) {
    schema[name] = schema[name].max(maxValue, translate(name, 'max'))
  }

  if (required) {
    schema[name] = schema[name].required(translate(name, 'required'))
  }

  // etc...
}
```